mem/buffer_slice.go

```diff
@@ -0,0 +1,155 @@
+package mem
+
+import (
+	"io"
+)
+
+// BufferSlice offers a means to represent data that spans one or more Buffer
+// instances.
+type BufferSlice []*Buffer
+
+// Len returns the sum of the length of all the Buffers in this slice.
+//
+// # Warning
+//
+// Invoking the built-in len on a BufferSlice will return the number of buffers
+// in the slice, and *not* the value returned by this function.
+func (s BufferSlice) Len() int {
+	var length int
+	for _, b := range s {
+		length += b.Len()
+	}
+	return length
+}
+
+// Ref returns a new BufferSlice containing a new reference of each Buffer in the
+// input slice.
+func (s BufferSlice) Ref() BufferSlice {
+	out := make(BufferSlice, len(s))
+	for i, b := range s {
+		out[i] = b.Ref()
+	}
+	return out
+}
+
+// Free invokes Buffer.Free on each Buffer in the slice.
+func (s BufferSlice) Free() {
+	for _, b := range s {
+		b.Free()
+	}
+}
+
+// CopyTo copies each of the underlying Buffer's data into the given buffer,
+// returning the number of bytes copied.
+func (s BufferSlice) CopyTo(out []byte) int {
+	off := 0
+	for _, b := range s {
+		off += copy(out[off:], b.ReadOnlyData())
+	}
+	return off
+}
+
+// Materialize concatenates all the underlying Buffer's data into a single
+// contiguous buffer using CopyTo.
+func (s BufferSlice) Materialize() []byte {
+	out := make([]byte, s.Len())
+	s.CopyTo(out)
+	return out
+}
+
+// LazyMaterialize functions like Materialize except that it writes the data to a
+// single Buffer pulled from the given BufferPool. As a special case, if the
+// input BufferSlice only actually has one Buffer, this function has nothing to
+// do and simply returns said Buffer, hence it being "lazy".
+func (s BufferSlice) LazyMaterialize(pool BufferPool) *Buffer {
+	if len(s) == 1 {
+		return s[0].Ref()
+	}
+	buf := pool.Get(s.Len())
+	s.CopyTo(buf)
+	return NewBuffer(buf, pool.Put)
+}
+
+// Reader returns a new Reader for the input slice.
+func (s BufferSlice) Reader() *Reader {
+	return &Reader{
+		data: s,
+		len:  s.Len(),
+	}
+}
+
+var _ io.ReadCloser = (*Reader)(nil)
+
+// Reader exposes a BufferSlice's data as an io.Reader, allowing it to interface
+// with other parts systems. It also provides an additional convenience method
+// Remaining which returns the number of unread bytes remaining in the slice.
+//
+// Note that reading data from the reader does not free the underlying buffers!
+// Only calling Close once all data is read will free the buffers.
```

Comment:
Should the `Reader` take a ref to all the underlying buffers instead? With the current approach, we will `panic` if `Free` is called on the `BufferSlice` *and* `Close` is called on the `Reader`. With the approach that I'm suggesting, both methods would have to be called, else it would leak the buffers. Did you pick this option over the other for some reason?

Reply:
From that comment I see that a reader will be used in the `loopyWriter` to read buffers from a `BufferSlice` inside of the `dataFrame`, and write it out via the framer. So, I get the point that we cannot free the buffers after reading them, because that would make it impossible to retry the write.

My suggestion is that:
- `Reader` takes a reference to the whole `BufferSlice`
- The `dataFrame` or any other entity that owns the `BufferSlice` also retains a reference
- Once the `Read` operation is complete, someone would have to call `Close` on the `Reader`
- And someone would also have to release the original reference to the `BufferSlice`

The reason I prefer this approach is that it ensures that the `BufferSlice` is not freed from under the `Reader`. Do you see any issues with this approach? Thanks.